06-r-and-sql.Rmd

```diff
@@ -0,0 +1,144 @@
+---------
+
+Interacting with databases through scripted languages can offer advantages over querying databases via a GUI interface.  A GUI interfaces for your database is easier to use and allows the quick viewing of adhoc queries.  Querying a database with a programatic intreface (in this case R, but it could be any language) are slightly more complicated.  However the trade-off is that data manipulations are preserved in the code.  Aggregations, summaries and other database operations are preserved.  Therefore those pre-analysis data manipulation steps are not lost.
+
+Connecting R to sqlite databases
+---------
+
+R can connect to databases through a number of libraries.  In our case we we will use [RSQLite](http://cran.r-project.org/web/packages/RSQLite/index.html) to connect to existing SQLite3 databases.  However you should be able to connect on almost any database in R via JDBC[http://cran.r-project.org/web/packages/RJDBC/index.html](http://cran.r-project.org/web/packages/RJDBC/index.html) or [ODBC](http://cran.r-project.org/web/packages/RODBC/index.html), or specific database libraries (such as we are doing, or [MySQL](http://cran.r-project.org/web/packages/RMySQL/index.html) ).  
+
+To begin these exercises we'll connect to the mammal database you've just created.
+
+```{r connect}
+library(RSQLite)
+## Set dbname and driver out of convenience
+myDB <- "~/Downloads/portal_mammals.sqlite"
+conn <- dbConnect(drv = SQLite(), dbname= myDB)
+```
+
+Now you're connected to your database, you can get information about the tables in the database.  While SQLite manager will provide this for you, you'll need to get it yourself with some basic SQL commands
+
+Basic queries
+----
+You can grab information about all tables like this:
+
+```{r Get Table information}
+dbGetQuery(conn,"SELECT type, tbl_name  FROM sqlite_master")
+```
+
+`RSQLite` has some functions that easily list basic information about your database and tables. Here you can see the types and names of fields and get a count of records.
+
+```{r table details}
+dbListTables(conn)
+dbListFields(conn,"surveys")
+dbGetQuery(conn,"SELECT count(*) FROM surveys")
+```
+
+Great!  That's all there is to it. You are now connected to your database and know something about the tables in it.  From here you can manipulate it the exact same way you can from SQLite Manager except in a scripted manner. Let's try some basic queries from the previous lesson.  Querying simply takes a connection to a database and query as inputs and returns a dataframe with the results.
+
+```{r basic queries}
+q <- 'SELECT DISTINCT year, species_id FROM surveys'
+result <-  dbGetQuery(conn,q)
+head(result)
+```
+
+
+
+>  **Challenge**
+>
+> Write a query that gets counts of genus by plot type. If this seems to advanaced, then get counts of genus by plot_id (eliminates of of the joins)
+
+*Answer*
+```{r genus by type}
+q <- "SELECT d.plot_type , c.genus, count(*)
+FROM
+(SELECT a.genus, b.plot
+FROM species a
```

Comment:
we don't cover table name aliasing in the lesson, but I think it's a good idea to demonstrate it here.